lib/stages/HFBTruncate.cpp

```diff
@@ -0,0 +1,141 @@
+#include "HFBTruncate.hpp"
+
+#include "Config.hpp"         // for Config
+#include "HFBFrameView.hpp"   // for HFBFrameView
+#include "StageFactory.hpp"   // for REGISTER_KOTEKAN_STAGE, StageMakerTemplate
+#include "buffer.h"           // for wait_for_full_frame, allocate_new_metadata_object, mark_fr...
+#include "kotekanLogging.hpp" // for DEBUG
+#include "truncate.hpp"       // for bit_truncate_float
+#include "visUtil.hpp"        // for cfloat
+
+#include "gsl-lite.hpp" // for span
+
+#include <atomic>      // for atomic_bool
+#include <cmath>       // for abs, sqrt
+#include <complex>     // for complex
+#include <cstdint>     // for int32_t
+#include <cstring>     // for memset, size_t
+#include <exception>   // for exception
+#include <functional>  // for _Bind_helper<>::type, bind, function
+#include <immintrin.h> // for _mm256_broadcast_ss, _mm256_div_ps, _mm256_loadu_ps, _mm25...
+#include <mm_malloc.h> // for _mm_free, _mm_malloc
+#include <regex>       // for match_results<>::_Base_type
+#include <stdexcept>   // for invalid_argument
+#include <vector>      // for vector
+
+
+using kotekan::bufferContainer;
+using kotekan::Config;
+using kotekan::Stage;
+
+REGISTER_KOTEKAN_STAGE(HFBTruncate);
+
+HFBTruncate::HFBTruncate(Config& config, const std::string& unique_name,
+                         bufferContainer& buffer_container) :
+    Stage(config, unique_name, buffer_container, std::bind(&HFBTruncate::main_thread, this)) {
+
+    // Fetch the buffers, register
+    in_buf = get_buffer("in_buf");
+    register_consumer(in_buf, unique_name.c_str());
+    out_buf = get_buffer("out_buf");
+    register_producer(out_buf, unique_name.c_str());
+
+    // Get truncation parameters from config
+    err_sq_lim = config.get<float>(unique_name, "err_sq_lim");
+    if (err_sq_lim < 0)
+        throw std::invalid_argument("HFBTruncate: config: err_sq_lim should"
+                                    " be positive (is "
+                                    + std::to_string(err_sq_lim) + ").");
+    w_prec = config.get<float>(unique_name, "weight_fixed_precision");
+    if (w_prec < 0)
+        throw std::invalid_argument("HFBTruncate: config: "
+                                    "weight_fixed_precision should be positive (is "
+                                    + std::to_string(w_prec) + ").");
+    hfb_prec = config.get<float>(unique_name, "data_fixed_precision");
+    if (hfb_prec < 0)
+        throw std::invalid_argument("HFBTruncate: config: "
+                                    "data_fixed_precision should be positive (is "
+                                    + std::to_string(hfb_prec) + ").");
```

Comment:
I'm not sure what the best way of exiting in this circumstance is. Should these be FATAL_ERROR's instead?